src/backend/distributed/operations/worker_split_copy_udf.c

```diff
@@ -0,0 +1,256 @@
+#include "utils/lsyscache.h"
+#include "utils/array.h"
+#include "utils/builtins.h"
+#include "distributed/listutils.h"
+#include "distributed/multi_executor.h"
+#include "distributed/worker_shard_copy.h"
+#include "distributed/intermediate_results.h"
+#include "distributed/citus_ruleutils.h"
+
+PG_FUNCTION_INFO_V1(worker_split_copy);
+
+typedef struct SplitCopyInfo
+{
+	uint64 destinationShardId;              /* destination shard id */
+	Datum destinationShardMinHashValue;     /* min hash value of destination shard */
+	Datum destinationShardMaxHashValue;     /* max hash value of destination shard */
+	uint32_t destinationShardNodeId;        /* node where split child shard is to be placed */
+} SplitCopyInfo;
+
+static void ParseSplitCopyInfoDatum(Datum splitCopyInfoDatum,
+									SplitCopyInfo **splitCopyInfo);
+static DestReceiver ** CreateShardCopyDestReceivers(EState *estate,
+													ShardInterval *
+													shardIntervalToSplitCopy,
+													List *splitCopyInfoList);
+static DestReceiver *  CreatePartitionedSplitCopyDestReceiver(EState *executor,
+															  ShardInterval *
+															  shardIntervalToSplitCopy,
+															  List *splitCopyInfoList);
+static void BuildMinMaxRangeArrays(List *splitCopyInfoList, ArrayType **minValueArray,
+								   ArrayType **maxValueArray);
+
+/*
+ * worker_split_copy(source_shard_id bigint, splitCopyInfo citus.split_copy_info[])
+ * UDF to split copy shard to list of destination shards.
+ * 'source_shard_id' : Source ShardId to split copy.
+ * 'splitCopyInfos'   : Array of Split Copy Info (destination_shard's id, min/max ranges and node_id)
+ */
+Datum
+worker_split_copy(PG_FUNCTION_ARGS)
+{
+	uint64 shardIdToSplitCopy = DatumGetUInt64(PG_GETARG_DATUM(0));
+	ShardInterval *shardIntervalToSplitCopy = LoadShardInterval(shardIdToSplitCopy);
+
+	ArrayType *splitCopyInfoArrayObject = PG_GETARG_ARRAYTYPE_P(1);
+	if (array_contains_nulls(splitCopyInfoArrayObject))
+	{
+		ereport(ERROR,
+				(errmsg("Shard Copy Info cannot have null values.")));
+	}
+
+	const int slice_ndim = 0;
+	ArrayMetaState *mState = NULL;
+	ArrayIterator copyInfo_iterator = array_create_iterator(splitCopyInfoArrayObject,
+															slice_ndim,
+															mState);
+	Datum copyInfoDatum = 0;
+	bool isnull = false;
+	List *splitCopyInfoList = NULL;
+	while (array_iterate(copyInfo_iterator, &copyInfoDatum, &isnull))
+	{
+		SplitCopyInfo *splitCopyInfo = NULL;
+		ParseSplitCopyInfoDatum(copyInfoDatum, &splitCopyInfo);
+
+		splitCopyInfoList = lappend(splitCopyInfoList, splitCopyInfo);
+	}
+
+	EState *executor = CreateExecutorState();
+	DestReceiver *splitCopyDestReceiver = CreatePartitionedSplitCopyDestReceiver(executor,
+																				 shardIntervalToSplitCopy,
+																				 splitCopyInfoList);
+
+	char *sourceShardToCopyName = generate_qualified_relation_name(
+		shardIntervalToSplitCopy->relationId);
+	AppendShardIdToName(&sourceShardToCopyName, shardIdToSplitCopy);
```

Comment:
I don't think this function knows how to work on a qualified name, which might be quoted. Would be good to add some tests with table names that have characters that require quoting.

Reply:
Added a test for this and it seems to work fine : https://github.com/citusdata/citus/pull/6029/files#diff-0162d076a0cf68ff1042c85cb2583ef38c69c4335c0ed970e702599aec036837.

Reply:
code is is still broken though :)

What happens is that for a table named "test-1" you get something like `SELECT * FROM public."test-1"_102010; ` . In that case postgres treats "_102010" as an alias for public."test-1". It's accidentally a valid query that queries the whole distributed table.

Reply:
Got it. Validate the bug with the test when I create 2 shards. Any existing helpers we use to solve this?

Reply:
Updated. Test now has two shards and verifies we only copy the right shard.